src/ipa_pc/mod.rs

```diff
@@ -736,6 +736,155 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
+                    / 
+                    &Polynomial::from_coefficients_vec(vec![
+                        (G::ScalarField::zero() - point), 
+                        G::ScalarField::one()
+                    ]);
+                
+                // h(X) = SUM( lambda^i * ((p_i(X) - y_i) / (X - x_i)) )
+                batch_polynomial += (cur_challenge, &polynomial);
+
+                // lambda^i
+                cur_challenge = opening_challenges(opening_challenge_counter);
+                opening_challenge_counter += 1;
+            }
+        }
+
+        // Commitment of the h(X) polynomial
+        let batch_commitment = Self::cm_commit(
+            ck.comm_key.as_slice(),
+            batch_polynomial.coeffs.as_slice(),
+            None,
+            None,
+        ).into_affine();
+
+        // TODO: When we will move to Sponge-based construction, we will absorb the batch_commitment only and then squeeze a new challenge
+        // Fresh random challenge x
+        let point= Self::compute_random_oracle_challenge(
+            &to_bytes![
+                batch_commitment,
+                points                
+            ]
+            .unwrap(),
+        );
+
+        // Values: v_i = p_i(x), where x is fresh random challenge
+        let mut batch_values = vec![];
+        for labeled_polynomial in labeled_polynomials.clone().into_iter() {
+            batch_values.push(labeled_polynomial.polynomial().evaluate(point));
+        }
+
+        // h(X) polynomial added to the set of polynomials for multi-poly single-point batching
+        let mut labeled_polynomials = labeled_polynomials.clone().into_iter().collect::<Vec<&'a LabeledPolynomial<G::ScalarField>>>();
+        let labeled_batch_polynomial = LabeledPolynomial::new(
+            format!("Batch"),
+            batch_polynomial,
+            None,
+            None
+        );
+        labeled_polynomials.push(&labeled_batch_polynomial);
+
+        // Commitment of h(X) polynomial added to the set of polynomials for multi-poly single-point batching
+        let mut commitments = commitments.clone().into_iter().collect::<Vec<&'a LabeledCommitment<Self::Commitment>>>();
+        let labeled_batch_commitment = LabeledCommitment::new(
+            format!("Batch"),
+            Commitment { comm: batch_commitment, shifted_comm: None },
+            None
+        );
+        commitments.push(&labeled_batch_commitment);
+
+        let mut rands = rands.clone().into_iter().collect::<Vec<&'a Self::Randomness>>();
+        let batch_randomness = Randomness::empty();
+        rands.push(&batch_randomness);
+
+        end_timer!(batch_time);
+
+        let opening_challenge = Self::compute_random_oracle_challenge(
+            &to_bytes![
+                batch_values,
+                batch_commitment,
+                point               
+            ]
+            .unwrap(),
+        );
+        let opening_challenges = |pow| opening_challenge.pow(&[pow]);
```

Comment:
Ok, here ``opening_challenges()`` is clearly considered as stateful function, and we take responsability to absorb/acknowledge the opening claim outside the function ``open_individual_opening_challenges()``. Furthermore, a separate (clearly stateless) random oracle already is implemented by ``compute_random_oracle_challenge()``.

Taking this into account, I would recommend to do the Fiat Shamir as follows:

- We initialize our state by hashing the claim of the (batched) opening proof, ``state = compute_random_oracle_challenge(all commitments, all points, and all claimed values)``.
- We initialize our ``opening_challenges()`` by setting it as the power function based on ``state``.
- We "squeeze" the challenges lambda_1, lambda_2, ..., lambda_n by simply calling ``opening_challenges()`` with counter ``i=1,2,...,n``.
- The new opening point x (the "fresh" one) is squeezed by absorbing the new batch poly and the last lambda_n, ``x= compute_random_oracle_challenge(lambda_n, Cbatchedpoly)``.